Beam/Include/Beam/Queries/AsyncDataStore.hpp

```diff
@@ -0,0 +1,215 @@
+
+namespace Beam::Queries {
+
+  /**
+   * Asynchronously writes to a data store.
+   * @param <D> The type of data store to buffer writes to.
+   * @param <E> The type of EvaluatorTranslator used for filtering values.
+   */
+  template<typename D, typename E =
+    typename GetTryDereferenceType<D>::EvaluatorTranslatorFilter>
+  class AsyncDataStore : private boost::noncopyable {
+    public:
+
+      /** The type of data store to buffer writes to. */
+      using DataStore = GetTryDereferenceType<D>;
+
+      /** The type of query used to load values. */
+      using Query = typename DataStore::Query;
+
+      /** The type of index used. */
+      using Index = typename DataStore::Index;
+
+      /** The type of value to store. */
+      using Value = typename DataStore::Value;
+
+      /** The SequencedValue to store. */
+      using SequencedValue = typename DataStore::SequencedValue;
+
+      /** The IndexedValue to store. */
+      using IndexedValue = typename DataStore::IndexedValue;
+
+      /** The type of EvaluatorTranslator used for filtering values. */
+      using EvaluatorTranslatorFilter = E;
+
+      /**
+       * Constructs an AsyncDataStore.
+       * @param dataStore Initializes the data store to buffer data to.
+       */
+      template<typename DS>
+      AsyncDataStore(DS&& dataStore);
+
+      ~AsyncDataStore();
+
+      std::vector<SequencedValue> Load(const Query& query);
+
+      void Store(const IndexedValue& value);
+
+      void Store(const std::vector<IndexedValue>& values);
+
+      void Open();
+
+      void Close();
+
+    private:
+      using ReserveDataStore = LocalDataStore<Query, Value,
+        EvaluatorTranslatorFilter>;
+      mutable boost::mutex m_mutex;
+      GetOptionalLocalPtr<D> m_dataStore;
+      bool m_flushPending;
+      std::shared_ptr<ReserveDataStore> m_currentDataStore;
+      std::shared_ptr<ReserveDataStore> m_flushedDataStore;
+      IO::OpenState m_openState;
+      RoutineTaskQueue m_tasks;
+
+      void Shutdown();
+      void Flush();
+      void TestFlush();
+  };
+
+  template<typename D, typename E>
+  template<typename DS>
+  AsyncDataStore<D, E>::AsyncDataStore(DS&& dataStore)
+    : m_dataStore(std::forward<DS>(dataStore)),
+      m_flushPending(false),
+      m_currentDataStore(std::make_shared<ReserveDataStore>()),
+      m_flushedDataStore(m_currentDataStore) {}
+  
+  template<typename D, typename E>
+  std::vector<typename AsyncDataStore<D, E>::SequencedValue>
+      AsyncDataStore<D, E>::Load(const Query& query) {
+    auto buffer = std::shared_ptr<ReserveDataStore>();
+    {
+      auto lock = boost::lock_guard(m_mutex);
+      buffer = m_flushedDataStore;
+    }
+    auto matches = std::vector<SequencedValue>();
+    if(query.GetSnapshotLimit().GetType() == SnapshotLimit::Type::HEAD) {
+      matches = m_dataStore->Load(query);
+    } else {
+      matches = buffer->Load(query);
+    }
+    if(static_cast<int>(matches.size()) < query.GetSnapshotLimit().GetSize()) {
```

Comment:
I have another concern about this one. You mentioned that usually the data added are new, but it is not strictly guaranteed, nor does the API restrict users to that.
Having said that, older data may be present in the buffer than those available in the main data store. This logic, however, relying on data in the buffer being always newer.
As an example of unexpected behavior, consider `BufferedDataStore` with a buffer size of 100. My main data store contains all the new data, but I have received a data feed with 100 old entries. I insert store them in my `BufferedDataStore`; those elements now fully occupy the buffer. I then attempt to load 100 newest entries from `BufferedDataStore`, but instead of getting the actual head of the principal data feed I get those values from the lateral feed.

It seems to me that I have to load `SnapshotLimit` items from every data store involved and then select a total of `SnapshotLimit` newest/oldest according to the query.

Reply:
Another option is to load newest/oldest entries one-by-one and merge on the fly, but it may be less efficient depending on the data store used. Would also have to pre-load `m_currentDataStore` anyway to avoid locking the mutex for the entire merge process.

Reply:
Sure, implement a solution based on what you think will best accomplish the goal. Every DataStore implementation is required to be synchronized. The BufferedDataStore is available as a guide to help you implement the AsyncDataStore, but it's not intended to be the actual solution, so don't worry too much about it. It might contain bugs, unexpected behavior, or other issues, none of which should be present in the AsyncDataStore.